warehouse/warehouse.go

```diff
@@ -447,13 +470,135 @@ func isDestInProgress(warehouse warehouseutils.WarehouseT) bool {
+	if lastExecTime, ok := lastExecMap[connectionString(warehouse)]; ok && time.Now().Unix()-lastExecTime < freqInS {
 		return true
 	}
+	return false
+}
+
+func setLastExec(warehouse warehouseutils.WarehouseT) {
+	lastExecMapLock.Lock()
+	defer lastExecMapLock.Unlock()
 	lastExecMap[connectionString(warehouse)] = time.Now().Unix()
+}
+
+// parseTimeToMinsOfDay returns minutes since start of day for a timestamp in format `15:30`
+// eg. parseTimeToMinsOfDay("02:30") -> 150
+func parseTimeToMinsOfDay(str string) int {
+	x := strings.Split(str, ":")
+	hrs, err := strconv.Atoi(x[0])
+	if err != nil {
+		panic(err)
+	}
+	mins, err := strconv.Atoi(x[1])
+	if err != nil {
+		panic(err)
+	}
+	return (hrs * 60) + mins
+}
+
+// scheduledTimes returns all possible start time
+// eg. Syncing every 3hrs starting at 13:00 (scheduled times: 13:00, 16:00, 19:00, 22:00, 01:00, 04:00, 07:00, 10:00)
+func scheduledTimes(syncFrequency, syncStartAt string) (times []int) {
+	if cachedTimes, ok := scheduledTimesCache[fmt.Sprintf(`%s-%s`, syncFrequency, syncStartAt)]; ok {
+		return cachedTimes
+	}
+	syncStartAtInMin := parseTimeToMinsOfDay(syncStartAt)
+	syncFrequencyInMin, _ := strconv.Atoi(syncFrequency)
+	times = append(times, syncStartAtInMin)
+	counter := 1
+	for {
+		mins := syncStartAtInMin + counter*syncFrequencyInMin
+		if mins >= 1440 {
+			break
+		}
+		times = append(times, mins)
+		counter++
+	}
+
+	prependTimes := []int{}
+	counter = 1
+	for {
+		mins := syncStartAtInMin - counter*syncFrequencyInMin
+		if mins < 0 {
+			break
+		}
+		prependTimes = append(prependTimes, mins)
+		counter++
+	}
+	times = append(funk.ReverseInt(prependTimes), times...)
+	scheduledTimesCache[fmt.Sprintf(`%s-%s`, syncFrequency, syncStartAt)] = times
+	return
+}
+
+func startOfDay(t time.Time) time.Time {
+	return time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, t.Location())
+}
+
+// getPrevScheduledTime returns closest previous scheduled time
+// eg. Syncing every 3hrs starting at 13:00 (scheduled times: 13:00, 16:00, 19:00, 22:00, 01:00, 04:00, 07:00, 10:00)
+// prev scheduled time for current time (eg. 18:00 -> 16:00 same day, 00:30 -> 22:00 prev day)
+func getPrevScheduledTime(syncFrequency, syncStartAt string) time.Time {
+	allStartTimes := scheduledTimes(syncFrequency, syncStartAt)
+
+	loc, _ := time.LoadLocation("UTC")
+	now := time.Now().In(loc)
+	// current time in minutes since start of day
+	currMins := now.Hour()*60 + now.Minute()
+
+	// get position where current time can fit in the sorted list of allallStartTimes
+	pos := 0
+	for idx, t := range allStartTimes {
+		if currMins > t {
+			continue
+		}
+		pos = idx - 1
+		break
+	}
+
+	// if current time is less than first start time in a day, take last start time in prev day
+	if pos < 0 {
+		return startOfDay(now).Add(time.Hour * time.Duration(-24)).Add(time.Minute * time.Duration(allStartTimes[len(allStartTimes)-1]))
+	}
+	return startOfDay(now).Add(time.Minute * time.Duration(allStartTimes[pos]))
+}
```

Comment:
I think there is a bug here: 
if current time is after last value of allStartTimes (e.g last value is 22:00 and current time is 23:00), pos variable will never be set based on idx, and will remain 0. This will result in adding first value's minutes to current start of day.

I suggest you create a small test for this function specifically, as it is easy to miss edge cases when times are close to day's starts/ends (I now I did have similar bugs in similar Blendo's code :)

Keep in mind, testing this function will require a small refactoring passing the current time as argument, instead of fixing it to time.Now()

Reply:
fixed that and added unit tests for the func